builtin/game/falling.lua

```diff
@@ -160,36 +160,71 @@ function nodeupdate_single(p, delay)
 					core.get_node_level(p_bottom) < core.get_node_max_level(p_bottom))) and
 				(not core.registered_nodes[n_bottom.name].walkable or
 					core.registered_nodes[n_bottom.name].buildable_to) then
-			if delay then
-				core.after(0.1, nodeupdate_single, p, false)
-			else
-				n.level = core.get_node_level(p)
-				core.remove_node(p)
-				spawn_falling_node(p, n)
-				nodeupdate(p)
-			end
+			n.level = core.get_node_level(p)
+			core.remove_node(p)
+			spawn_falling_node(p, n)
+			return true
 		end
 	end
 
 	if core.get_item_group(n.name, "attached_node") ~= 0 then
 		if not check_attached_node(p, n) then
 			drop_attached_node(p)
-			nodeupdate(p)
+			return true
 		end
 	end
+
+	return false
 end
 
+-- this table is specifically ordered.
+-- we don't walk diagonals, only our direct neighbors, and self
+-- down first as likely case, but always before self. same with sides.
+-- up must come last
+local nodeupdate_neighbors = {
+	{x = 0, y = -1, z = 0},
+	{x = -1, y = 0, z = 0},
+	{x = 1, y = 0, z = 0},
+	{x = 0, y = 0, z = 1},
+	{x = 0, y = 0, z = -1},
+	{x = 0, y = 0, z = 0},
+	{x = 0, y = 1, z = 0},
+}
+
 function nodeupdate(p, delay)
 	-- Round p to prevent falling entities to get stuck
 	p = vector.round(p)
 
-	for x = -1, 1 do
-	for y = -1, 1 do
-	for z = -1, 1 do
-		local d = vector.new(x, y, z)
-		nodeupdate_single(vector.add(p, d), delay or not (x == 0 and y == 0 and z == 0))
-	end
-	end
+	-- we make a stack, and manually maintain size for performance
+	local s, n = {}, 0
```

Comment:
This is just a personal preference thing, but could you split this compound initialization on separate lines?  I got kind of confused here for a second because I'm not as used to Lua.

```
local s = {}
local n = 0
```